File: core-rs/src/crypto.rs
//! 会话密码学：HKDF-SHA256 派生 + AES-256-GCM 信封加密。
//!
//! 密钥分离：每个方向（client→server / server→client）独立密钥；
//! nonce 为 12B：8B 大端计数器 + 4B 零前缀；AAD 参与认证。

use aes_gcm::aead::consts::U12;
use aes_gcm::aead::generic_array::GenericArray;
use aes_gcm::aead::{Aead, KeyInit, Payload};
use aes_gcm::{Aes256Gcm, Key};
use hkdf::Hkdf;
use rand::RngCore;
use sha2::Sha256;
use subtle::ConstantTimeEq;
use thiserror::Error;
use zeroize::Zeroize;

/// RFC 5869 HKDF-SHA256：extract + expand。
pub fn hkdf_sha256(salt: &[u8], ikm: &[u8], info: &[u8], okm: &mut [u8]) {
    let hk = Hkdf::<Sha256>::new(if salt.is_empty() { None } else { Some(salt) }, ikm);
    hk.expand(info, okm).expect("hkdf expand overflow");
}

/// 双向会话密钥集（各 32B）。
#[derive(Debug, Clone)]
pub struct SessionKeys {
    pub client_to_server: [u8; 32],
    pub server_to_client: [u8; 32],
}

impl SessionKeys {
    /// 从 PSK（配对协商的 32B 共享密钥）与双方随机数派生。
    pub fn derive(
        psk: &[u8; 32],
        client_random: &[u8; 32],
        server_random: &[u8; 32],
    ) -> SessionKeys {
        let hk = Hkdf::<Sha256>::new(
            Some(&[client_random.as_slice(), server_random.as_slice()].concat()),
            psk,
        );
        let mut c2s = [0u8; 32];
        let mut s2c = [0u8; 32];
        hk.expand(b"tvremote c2s key", &mut c2s).expect("hmac length");
        hk.expand(b"tvremote s2c key", &mut s2c).expect("hmac length");
        SessionKeys {
            client_to_server: c2s,
            server_to_client: s2c,
        }
    }
}

#[derive(Debug, Error)]
pub enum CryptoError {
    #[error("payload too large: {0}")]
    TooLarge(usize),
    #[error("authentication failed")]
    AuthFailed,
    #[error("nonce exhausted")]
    NonceExhausted,
    #[error("crypto error: {0}")]
    Other(String),
}

/// 单方向加密器：持有密钥与发送计数器。
pub struct DirectionCipher {
    cipher: Aes256Gcm,
    send_counter: u64,
}

impl DirectionCipher {
    pub fn new(key: [u8; 32]) -> Self {
        DirectionCipher {
            cipher: Aes256Gcm::new(Key::<Aes256Gcm>::from_slice(&key)),
            send_counter: 0,
        }
    }

    /// 12B nonce：4B 零前缀 + 8B 大端计数器。
    fn nonce(counter: u64) -> GenericArray<u8, U12> {
        let mut n = [0u8; 12];
        n[4..].copy_from_slice(&counter.to_be_bytes());
        *GenericArray::from_slice(&n)
    }

    /// 加密（AAD 参与认证）。返回 ciphertext || tag。
    pub fn seal(&mut self, plaintext: &[u8], aad: &[u8]) -> Result<Vec<u8>, CryptoError> {
        let nonce = Self::nonce(self.send_counter);
        self.send_counter = self
            .send_counter
            .checked_add(1)
            .ok_or(CryptoError::NonceExhausted)?;
        self.cipher
            .encrypt(&nonce, Payload { msg: plaintext, aad })
            .map_err(|_| CryptoError::AuthFailed)
    }

    /// 解密（计数器由调用方按对端序号维护）。
    pub fn open(
        &self,
        ciphertext: &[u8],
        counter: u64,
        aad: &[u8],
    ) -> Result<Vec<u8>, CryptoError> {
        let nonce = Self::nonce(counter);
        self.cipher
            .decrypt(&nonce, Payload { msg: ciphertext, aad })
            .map_err(|_| CryptoError::AuthFailed)
    }
}

/// 常数时间比较（等长才可能相等）。
pub fn ct_eq(a: &[u8], b: &[u8]) -> bool {
    a.len() == b.len() && bool::from(a.ct_eq(b))
}

/// 加密随机字节。
pub fn random_bytes(buf: &mut [u8]) {
    rand::rngs::OsRng.fill_bytes(buf);
}

/// 内存清零（密钥生命周期结束时调用）。
pub fn zeroize_key(key: &mut [u8]) {
    key.zeroize();
}

#[cfg(test)]
mod tests {
    use super::*;
    use hex_literal::hex;

    #[test]
    fn hkdf_rfc5869_test_case_1() {
        // RFC 5869 A.1: SHA-256 基础向量
        let ikm = hex!("0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b");
        let salt = hex!("000102030405060708090a0b0c");
        let info = hex!("f0f1f2f3f4f5f6f7f8f9");
        let mut okm = [0u8; 42];
        hkdf_sha256(&salt, &ikm, &info, &mut okm);
        let expect = hex!(
            "3cb25f25faacd57a90434f64d0362f2a
             2d2d0a90cf1a5a4c5db02d56ecc4c5bf
             34007208d5b887185865"
        );
        assert_eq!(okm, expect);
    }

    #[test]
    fn seal_open_roundtrip() {
        let mut a = DirectionCipher::new([7u8; 32]);
        let b = DirectionCipher::new([7u8; 32]);
        let ct = a.seal(b"hello", b"aad1").unwrap();
        let pt = b.open(&ct, 0, b"aad1").unwrap();
        assert_eq!(pt, b"hello");
    }

    #[test]
    fn aad_mismatch_fails() {
        let mut a = DirectionCipher::new([1u8; 32]);
        let b = DirectionCipher::new([1u8; 32]);
        let ct = a.seal(b"x", b"aad1").unwrap();
        assert!(b.open(&ct, 0, b"aad2").is_err());
    }

    #[test]
    fn replay_with_wrong_counter_fails() {
        let mut a = DirectionCipher::new([3u8; 32]);
        let b = DirectionCipher::new([3u8; 32]);
        let ct = a.seal(b"m", b"").unwrap();
        assert!(b.open(&ct, 5, b"").is_err());
    }

    #[test]
    fn ct_eq_edge() {
        assert!(ct_eq(&[], &[]));
        assert!(!ct_eq(&[1], &[2]));
        assert!(!ct_eq(&[1], &[1, 2]));
    }
}

File: miniprograms/core/src/crypto.ts
/**
 * 会话加密：HKDF-SHA256 派生双向密钥 + AES-256-GCM（纯 JS，经 @noble/*）。
 *
 * 算法与 Kotlin `DebugSessionCrypto` / Rust `crypto.rs` 逐条对齐：
 * - salt = client_random || server_random；IKM = psk（32B）
 * - info = "tvremote c2s key" / "tvremote s2c key"，各出 32B
 * - nonce = 12B：4B 零前缀 + 8B 大端 counter
 * - 密文布局：counter(8B) || ciphertext || tag(16B)
 * - counter 与 extraAad 并入加密明文体（认证绑定；counter 从 1 起，0 为防重放哨兵）
 */

import { hkdf } from "@noble/hashes/hkdf";
import { sha256 } from "@noble/hashes/sha256";
import { gcm } from "@noble/ciphers/aes";
import { readCounterBigEndian, writeCounterBigEndian } from "./frame.js";
import { utf8Encode } from "./text.js";

export const SECRET_BYTES = 32;
export const RANDOM_BYTES = 32;
export const COUNTER_BYTES = 8;
export const GCM_TAG_BYTES = 16;

export interface SessionKeys {
  clientToServer: Uint8Array;
  serverToClient: Uint8Array;
}

/** 双向密钥派生（与 Kotlin/Rust 双端黄金向量一致）。 */
export function deriveSessionKeys(
  psk: Uint8Array,
  clientRandom: Uint8Array,
  serverRandom: Uint8Array,
): SessionKeys {
  if (psk.length !== SECRET_BYTES) throw new Error("psk must be 32 bytes");
  if (clientRandom.length !== RANDOM_BYTES || serverRandom.length !== RANDOM_BYTES) {
    throw new Error("randoms must be 32 bytes");
  }
  const salt = new Uint8Array(clientRandom.length + serverRandom.length);
  salt.set(clientRandom, 0);
  salt.set(serverRandom, clientRandom.length);
  return {
    clientToServer: hkdf(sha256, psk, salt, utf8Encode("tvremote c2s key"), SECRET_BYTES),
    serverToClient: hkdf(sha256, psk, salt, utf8Encode("tvremote s2c key"), SECRET_BYTES),
  };
}

/** RFC 5869 HKDF-SHA256（与 Kotlin DebugSessionCrypto.hkdfSha256 同实现语义）。 */
export function hkdfSha256(salt: Uint8Array, ikm: Uint8Array, info: Uint8Array, outLength: number): Uint8Array {
  return hkdf(sha256, ikm, salt, info, outLength);
}

/** nonce = 12B：4B 零前缀 + 8B 大端 counter。 */
export function nonceFor(counter: bigint): Uint8Array {
  const nonce = new Uint8Array(12);
  nonce.set(writeCounterBigEndian(counter, COUNTER_BYTES), 4);
  return nonce;
}

export interface SealResult {
  /** counter(8B) || ciphertext || tag(16B) */
  envelope: Uint8Array;
  counter: bigint;
}

/** 单方向加密封装：发送计数器内部递增；counter 与 extraAad 并入加密明文体。 */
export class DirectionCipher {
  private readonly key: Uint8Array;
  private sendCounter = 1n;

  constructor(key: Uint8Array) {
    if (key.length !== SECRET_BYTES) throw new Error("session key must be 32 bytes");
    this.key = key;
  }

  seal(plaintext: Uint8Array, extraAad: Uint8Array = new Uint8Array(0)): SealResult {
    if (this.sendCounter === 0xfffffffffffffffn) throw new Error("nonce exhausted");
    const counter = this.sendCounter;
    this.sendCounter += 1n;
    const body = new Uint8Array(COUNTER_BYTES + extraAad.length + plaintext.length);
    body.set(writeCounterBigEndian(counter, COUNTER_BYTES), 0);
    body.set(extraAad, COUNTER_BYTES);
    body.set(plaintext, COUNTER_BYTES + extraAad.length);
    const cipher = gcm(this.key, nonceFor(counter));
    const encrypted = cipher.encrypt(body);
    const envelope = new Uint8Array(COUNTER_BYTES + encrypted.length);
    envelope.set(writeCounterBigEndian(counter, COUNTER_BYTES), 0);
    envelope.set(encrypted, COUNTER_BYTES);
    return { envelope, counter };
  }

  /** 解密（对端计数器由调用方经防重放窗口校验后传入；内部校验加密体中的 counter/AAD）。 */
  open(envelope: Uint8Array, counter: bigint, extraAad: Uint8Array = new Uint8Array(0)): Uint8Array {
    if (envelope.length <= COUNTER_BYTES + GCM_TAG_BYTES) {
      throw new Error("ciphertext too short");
    }
    const embedded = readCounterBigEndian(envelope.slice(0, COUNTER_BYTES));
    if (embedded !== counter) throw new Error("counter mismatch");
    const cipher = gcm(this.key, nonceFor(counter));
    const body = cipher.decrypt(envelope.slice(COUNTER_BYTES));
    if (body.length < COUNTER_BYTES + extraAad.length) {
      throw new Error("decrypted body too short");
    }
    const expected = readCounterBigEndian(body.slice(0, COUNTER_BYTES));
    if (expected !== counter) throw new Error("inner counter mismatch");
    for (let i = 0; i < extraAad.length; i += 1) {
      if (body[COUNTER_BYTES + i] !== extraAad[i]) throw new Error("extra aad mismatch");
    }
    return body.slice(COUNTER_BYTES + extraAad.length);
  }
}

export { randomBytes, setRandomSource } from "./random.js";
